ProjFS.Mac/PrjFSKextTests/VirtualizationRootsTests.mm

```diff
@@ -264,4 +271,119 @@ - (void)testRegisterProviderForPath_ExistingRecycledRoot
+
+    PrjFSVirtualizationRootXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    repoRootVnode->xattrs.insert(make_pair(PrjFSVirtualizationRootXAttrName, rootXattrData));
+
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, testFileVnode.get(), self->dummyVFSContext);
+
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(foundRoot));
+    
+    if (VirtualizationRoot_IsValidRootHandle(foundRoot))
+    {
+        XCTAssertEqual(s_virtualizationRoots[foundRoot].rootInode, repoRootVnode->GetInode());
+        
+        repoRootVnode->StartRecycling();
+        
+        shared_ptr<vnode> newRepoRootVnode = self->testMountPoint->CreateVnode(repoPath, VnodeCreationProperties { .inode = repoRootVnode->GetInode(), .type = VDIR });
+        
+        shared_ptr<vnode> otherTestFileVnode = self->testMountPoint->CreateVnodeTree(otherFilePath);
+        VirtualizationRootHandle foundRootForOther = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, otherTestFileVnode.get(), self->dummyVFSContext);
+        
+        XCTAssertEqual(foundRootForOther, foundRoot, "Despite recycling the repo root vnode, we should end up with the same handle");
+        XCTAssertEqual(s_virtualizationRoots[foundRootForOther].rootVNode, newRepoRootVnode.get(), "Vnode should be refreshed");
+        XCTAssertEqual(s_virtualizationRoots[foundRootForOther].rootVNodeVid, newRepoRootVnode->GetVid(), "Vnode VID should be refreshed");
+    }
+}
+
+- (void)testFindForVnode_DirectoryInUndetectedRoot
+{
+    const char* repoPath = "/Users/test/code/Repo";
+    const char* subdirPath = "/Users/test/code/Repo/some/nested/subdir";
+    
+    shared_ptr<vnode> repoRootVnode = self->testMountPoint->CreateVnodeTree(repoPath, VDIR);
+    shared_ptr<vnode> testSubdirVnode = self->testMountPoint->CreateVnodeTree(subdirPath, VDIR);
+
+    XCTAssertTrue(vnode_isdir(testSubdirVnode.get()));
+
+    PrjFSVirtualizationRootXAttrData rootXattr = {};
+    vector<uint8_t> rootXattrData(sizeof(rootXattr), 0x00);
+    memcpy(rootXattrData.data(), &rootXattr, rootXattrData.size());
+    repoRootVnode->xattrs.insert(make_pair(PrjFSVirtualizationRootXAttrName, rootXattrData));
+
+    VirtualizationRootHandle foundRoot = VirtualizationRoot_FindForVnode(&self->dummyTracer, PrjFSPerfCounter_VnodeOp_FindRoot, PrjFSPerfCounter_VnodeOp_FindRoot_Iteration, testSubdirVnode.get(), self->dummyVFSContext);
+    
+    XCTAssertTrue(VirtualizationRoot_IsValidRootHandle(foundRoot));
+    if (VirtualizationRoot_IsValidRootHandle(foundRoot))
+    {
+        XCTAssertEqual(s_virtualizationRoots[foundRoot].rootInode, repoRootVnode->GetInode());
```

Comment:
Should we also double check the rootInode on the File version of this test? Should we do that on any other tests?

Reply:
The recycling test, which uses a file, [also checks this](https://github.com/Microsoft/VFSForGit/pull/908/files#diff-16401a67e162da73f89646a0727af85dR349) - I think that pretty much covers our bases. I'm not sure there's much value in asserting the kitchen sink in each test, my philosophy here is to stick to what's necessary to make the tests easier to read - when looking at the test I'd like it to be reasonably clear what failure case(s) it's protecting us from, which gets muddied when we test all possible invariants.

Reply:
That makes sense to me!  

For my understanding why do we need the check here and not in testFindForVnode_FileInUndetectedRoot?  Or is it just a matter of this one does the check so the other one shouldn't need to verify it.

Reply:
* `testFindForVnode_FileInUndetectedRoot` Generally checks that we can discover a previously unknown root by walking up the directory hierarchy and encountering the root, which has a special xattr.
 * `testFindForVnode_DetectRootRecycleThenFindRootForOtherFile` specifically tests against a regression to #797 - that bug manifested itself as the wrong inode being assigned to a root when discovering a previously unknown root via the xattr. So that's why that test checks the inode is correct. Additionally it checks that the inode-based detection works after the original root vnode was recycled. (Which is the most common situation in which #797 actually became noticeable.)
 * `testFindForVnode_DirectoryInUndetectedRoot` Is kind of a variation on parts of both of the above tests with a directory vnode rather than a regular file. It's specifically designed to cover the `if (vnode_isdir(vnode))` branch in `VirtualizationRoot_FindForVnode`. We probably don't *need* to check the inode here (#797 affected directories and regular files equally), but it doesn't hurt either, just in case taking that branch (which was newly introduced with the #797 fix) had any effect on the inode in particular.

If anything, `testFindForVnode_FileInUndetectedRoot` is superfluous, because the other two tests would fail for anything that might cause that test to fail. But I like having tests of escalating complexity for a lack of better term for it. Test some really basic scenarios as an initial sanity check, then test some very specific failure scenarios or variations.

Reply:
I see how `testFindForVnode_DirectoryInUndetectedRoot ` is a variation of `testFindForVnode_FileInUndetectedRoot`, but I'm not sure what part of it is a variation of `testFindForVnode_DetectRootRecycleThenFindRootForOtherFile`.

Is it because of this check?

```
XCTAssertEqual(s_virtualizationRoots[foundRoot].rootInode, repoRootVnode->GetInode());
```

Or is it that in both of the above tests we're discovering the root by its xattr?

(I think the test looks good, my question is just about how the test is related to `testFindForVnode_DetectRootRecycleThenFindRootForOtherFile`).

Reply:
> Is it because of this check?

Yes, like `testFindForVnode_DetectRootRecycleThenFindRootForOtherFile` it checks that `VirtualizationRoot_FindForVnode()` sets the correct inode in the detected root, but it checks this on the `if (vnode_isdir(vnode))` path of the function. I didn't think the rest of `testFindForVnode_DetectRootRecycleThenFindRootForOtherFile` (the recycling bit) really warranted repeating.